vertical-pod-autoscaler/pkg/admission-controller/logic/recommendation_provider.go

```diff
@@ -61,17 +64,36 @@ func NewRecommendationProvider(vpaLister vpa_lister.VerticalPodAutoscalerLister,
 	}
 }
 
-func limitMatchesRequest(limit, request *resource.Quantity) bool {
-	// Limit not set, it cannot match the request.
-	if limit == nil || limit.Value() == 0 {
-		return false
+func limitMatchesRequest(originalLimit, originalRequest, recommendedRequest *resource.Quantity) *resource.Quantity {
+	// originalLimit not set, don't set limit.
+	if originalLimit == nil || originalLimit.Value() == 0 {
+		return nil
+	}
+	// originalLimit set but originalRequest not set - K8s will treat the pod as if they were equal,
+	// recommend limit equal to request
+	if originalRequest == nil || originalRequest.Value() == 0 {
+		return recommendedRequest
 	}
-	// Limit set but request not set - K8s will treat the pod as if they were equal.
-	if request == nil || request.Value() == 0 {
-		return true
+	// originalLimit and originalRequest are set. If they are equal recommend limit equal to request.
+	if originalRequest.MilliValue() == originalLimit.MilliValue() {
+		return recommendedRequest
 	}
-	// Limit and request are set. Check if they are equal.
-	return request.Value() == limit.Value()
+
+	// Input and output milli values should fit in int64 but intermediate values might be bigger.
+	originalMilliRequest := big.NewInt(originalRequest.MilliValue())
```

Comment:
Wouldn't it be safer (without risking values > maxint) to compute actual ratio and then scale by ratio?

Reply:
I seem to recall that the k8s resource types have some nifty Scale() function (I may be wrong though)

Reply:
> Wouldn't it be safer (without risking values > maxint) to compute actual ratio and then scale by ratio?

It wouldn't be safer - user can still set request to 1 millibyte and limit to 1GB (for limit/request ratio of about 2^40). If VPA recommends limit of 1 GB limit recommendation will be about 2^80 milli bytes (2^30 bytes request * 2^10 (converting to milli units) * 2^40 ratio), well out of int64 range, regardless how we calculate the value.

But I think it doesn't really matter - we can't really give more than 2^64 bytes to anything anyways.

Reply:
> I seem to recall that the k8s resource types have some nifty Scale() function (I may be wrong though)

[I can't find it](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/vendor/k8s.io/apimachinery/pkg/api/resource/quantity.go).

Reply:
How about we calculate request / limit and if that is < epsilon we decide that this is an error, and actually not move the limit at all (instead of setting to infinity)?

Reply:
Setting to infinity is our best guess at users intent here. I added an annotation as you suggested.